app/models/concerns/unparsed_run.rb

```diff
@@ -13,154 +18,286 @@ def parse_into_db
+      # .with will clean up the iterator when we are done
+      run.metadata.variables.with do |iter|
+        until (variable = iter.next).nil?
+          run_data[:metadata][:variables][variable.name] = variable.value
+        end
+      end
 
-      RunHistory.import(histories.map do |history|
-        RunHistory.new(
-          run_id: id,
+      # Retrieve total playtime, and both sum of bests
+      # .with will clean up any objects that will not be GC'd
+      Parser::LiveSplitCore::Analysis.calculate_total_playtime_for_run(run)
+                                     .with { |analysis| run_data[:total_playtime_ms] = analysis.total_seconds * 1000 }
+      analysis = [
+        Parser::LiveSplitCore::Analysis.calculate_sum_of_best(run, false, false, 0),
+        Parser::LiveSplitCore::Analysis.calculate_sum_of_best(run, false, false, 1)
+      ]
+      run_data[:realtime_sum_of_best_ms] = analysis[0].present? ? analysis[0].total_seconds * 1000 : 0
+      run_data[:gametime_sum_of_best_ms] = analysis[1].present? ? analysis[1].total_seconds * 1000 : 0
+      # See https://github.com/glacials/splits-io/pull/474/files#r241242051
+      analysis.compact.each(&:dispose)
+
+      # Set run duration
+      last_seg = run.segment(run.len - 1)
+      run_data[:realtime_duration_ms] = last_seg.personal_best_split_time.real_time&.total_seconds&.*(1000) || 0
+      run_data[:gametime_duration_ms] = last_seg.personal_best_split_time.game_time&.total_seconds&.*(1000) || 0
+
+      run_data[:default_timing] = Run::REAL
+      if run_data[:realtime_duration_ms].zero? && run_data[:gametime_duration_ms].positive?
+        run_data[:default_timing] = Run::GAME
+      end
 
-          attempt_number:       history[:attempt_number],
-          realtime_duration_ms: zero_to_nil(history[:realtime_duration_ms]),
-          gametime_duration_ms: zero_to_nil(history[:gametime_duration_ms]),
+      run_data
+    end
 
-          started_at: history[:started_at],
-          ended_at:   history[:ended_at]
+    def process_run_history(run)
+      container = []
+
+      (0...run.attempt_history_len).each do |i|
+        attempt = run.attempt_history_index(i)
+        dates = [attempt.started, attempt.ended]
+
+        history = RunHistory.new(
+          run_id:               id,
+          attempt_number:       attempt.index.to_i,
+          realtime_duration_ms: attempt.time.real_time&.total_seconds&.*(1000) || nil,
+          gametime_duration_ms: attempt.time.game_time&.total_seconds&.*(1000) || nil,
+          pause_duration_ms:    attempt.pause_time&.total_seconds&.*(1000) || 0,
+          started_at:           dates[0] && DateTime.parse(dates[0].to_rfc3339),
+          ended_at:             dates[1] && DateTime.parse(dates[1].to_rfc3339)
         )
-      end)
+
+        # See https://github.com/glacials/splits-io/pull/474/files#r241242051
+        dates.compact.each(&:dispose)
+
+        container << history
+        next if container.size < MAX_CONTAINER_SIZE
```

Comment:
Do you know a better way to do this?  It seems hacky to check this and then also have to do it outside the loop in case we never hit the limit.

Reply:
Yeah see [`each_slice`][1] :D e.g.

```ruby
(0...run.attempt_history_len).each_slice(MAX_CONTAINER_SIZE) do |indexes|
  write_run_histories(indexes.map do |i|
    attempt = run.attempt_history_index(i)
    RunHistory.new(
      # ...
    )
  end)
end
```

[1]: https://apidock.com/ruby/Enumerable/each_slice